From b418fd5857bd227e38bc6476a7291808b0798b89 Mon Sep 17 00:00:00 2001
From: Mahesh Mahadevan <r9aadq@freescale.com>
Date: Mon, 10 Mar 2008 14:38:05 -0500
Subject: [PATCH] ENGR00056532 Support SDMA Transfer to FIFO Memory Device

Add Support to do memory to memory transfers between a FIFO Memory Device
and SDRAM

Signed-off-by: Mahesh Mahadevan <r9aadq@freescale.com>
---
 arch/arm/mach-mx3/devices.c       |    2 ++
 arch/arm/mach-mx3/dma.c           |   16 +++++++++++++++-
 arch/arm/plat-mxc/sdma/dma_sdma.c |    9 ++++++++-
 arch/arm/plat-mxc/sdma/sdma.c     |    2 ++
 include/asm-arm/arch-mxc/dma.h    |    3 ++-
 include/asm-arm/arch-mxc/mx31.h   |    5 +++--
 include/asm-arm/arch-mxc/sdma.h   |    9 +++++++--
 7 files changed, 39 insertions(+), 7 deletions(-)

diff --git a/arch/arm/mach-mx3/devices.c b/arch/arm/mach-mx3/devices.c
index e983bd5..559c797 100644
--- a/arch/arm/mach-mx3/devices.c
+++ b/arch/arm/mach-mx3/devices.c
@@ -81,6 +81,8 @@ void mxc_sdma_get_script_info(sdma_script_start_addrs * sdma_script_addr)
 		sdma_script_addr->mxc_sdma_app_2_mcu_addr =
 		    app_2_mcu_patched_ADDR_2;
 		sdma_script_addr->mxc_sdma_ap_2_ap_addr = ap_2_ap_ADDR_2;
+		sdma_script_addr->mxc_sdma_ap_2_ap_fixed_addr =
+		    ap_2_ap_fixed_addr_ADDR_2;
 		sdma_script_addr->mxc_sdma_ap_2_bp_addr = ap_2_bp_ADDR_2;
 		sdma_script_addr->mxc_sdma_bp_2_ap_addr = bp_2_ap_ADDR_2;
 		sdma_script_addr->mxc_sdma_loopback_on_dsp_side_addr = -1;
diff --git a/arch/arm/mach-mx3/dma.c b/arch/arm/mach-mx3/dma.c
index d2fa288..7499066 100644
--- a/arch/arm/mach-mx3/dma.c
+++ b/arch/arm/mach-mx3/dma.c
@@ -1,5 +1,5 @@
 /*
- *  Copyright 2007 Freescale Semiconductor, Inc. All Rights Reserved.
+ *  Copyright 2007-2008 Freescale Semiconductor, Inc. All Rights Reserved.
  */
 
 /*
@@ -615,6 +615,19 @@ static mxc_sdma_channel_params_t mxc_sdma_memory_params = {
 	.chnl_priority = MXC_SDMA_DEFAULT_PRIORITY,
 };
 
+static mxc_sdma_channel_params_t mxc_sdma_fifo_memory_params = {
+	.chnl_params = {
+			.peripheral_type = FIFO_MEMORY,
+			.per_address = MXC_FIFO_MEM_DEST_FIXED,
+			.transfer_type = emi_2_emi,
+			.bd_number = 32,
+			.word_size = TRANSFER_32BIT,
+			.event_id = 0,
+			},
+	.channel_num = MXC_DMA_CHANNEL_FIFO_MEMORY,
+	.chnl_priority = MXC_SDMA_DEFAULT_PRIORITY,
+};
+
 static mxc_sdma_channel_params_t mxc_sdma_ata_rx_params = {
 	.chnl_params = {
 			.watermark_level = MXC_IDE_DMA_WATERMARK,
@@ -686,6 +699,7 @@ static mxc_sdma_info_entry_t mxc_sdma_active_dma_info[] = {
 	{MXC_DMA_FIR_RX, &mxc_sdma_fir_rx_params},
 	{MXC_DMA_FIR_TX, &mxc_sdma_fir_tx_params},
 	{MXC_DMA_MEMORY, &mxc_sdma_memory_params},
+	{MXC_DMA_FIFO_MEMORY, &mxc_sdma_fifo_memory_params},
 	{MXC_DMA_ATA_RX, &mxc_sdma_ata_rx_params},
 	{MXC_DMA_ATA_TX, &mxc_sdma_ata_tx_params},
 };
diff --git a/arch/arm/plat-mxc/sdma/dma_sdma.c b/arch/arm/plat-mxc/sdma/dma_sdma.c
index 89fea14..bf0bd5e 100644
--- a/arch/arm/plat-mxc/sdma/dma_sdma.c
+++ b/arch/arm/plat-mxc/sdma/dma_sdma.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2004-2007 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright 2004-2008 Freescale Semiconductor, Inc. All Rights Reserved.
  */
 
 /*
@@ -322,6 +322,13 @@ int mxc_dma_config(int channel_num, mxc_dma_requestbuf_t * dma_buf,
 			} else {
 				chnl_param.transfer_type = emi_2_dsp;
 			}
+		} else if (chnl_param.peripheral_type == FIFO_MEMORY) {
+			if (mode == MXC_DMA_MODE_READ) {
+				chnl_param.per_address = MXC_FIFO_MEM_SRC_FIXED;
+			} else {
+				chnl_param.per_address =
+				    MXC_FIFO_MEM_DEST_FIXED;
+			}
 		} else {
 			if (mode == MXC_DMA_MODE_READ) {
 				chnl_param.transfer_type = per_2_emi;
diff --git a/arch/arm/plat-mxc/sdma/sdma.c b/arch/arm/plat-mxc/sdma/sdma.c
index d21c6a8..c9d52ba 100644
--- a/arch/arm/plat-mxc/sdma/sdma.c
+++ b/arch/arm/plat-mxc/sdma/sdma.c
@@ -359,6 +359,8 @@ static unsigned short sdma_get_pc(sdma_periphT peripheral_type,
 		default:
 			res = -EINVAL;
 		}
+	} else if (peripheral_type == FIFO_MEMORY) {
+		res = sdma_script_addrs.mxc_sdma_ap_2_ap_fixed_addr;
 	}
 
 	if (res < 0) {
diff --git a/include/asm-arm/arch-mxc/dma.h b/include/asm-arm/arch-mxc/dma.h
index 41997e6..496718f 100644
--- a/include/asm-arm/arch-mxc/dma.h
+++ b/include/asm-arm/arch-mxc/dma.h
@@ -1,5 +1,5 @@
 /*
- * Copyright 2004-2007 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright 2004-2008 Freescale Semiconductor, Inc. All Rights Reserved.
  */
 
 /*
@@ -77,6 +77,7 @@ typedef enum mxc_dma_device {
 	MXC_DMA_ATA_RX,
 	MXC_DMA_ATA_TX,
 	MXC_DMA_MEMORY,
+	MXC_DMA_FIFO_MEMORY,
 	MXC_DMA_DSP_PACKET_DATA0_RD,
 	MXC_DMA_DSP_PACKET_DATA0_WR,
 	MXC_DMA_DSP_PACKET_DATA1_RD,
diff --git a/include/asm-arm/arch-mxc/mx31.h b/include/asm-arm/arch-mxc/mx31.h
index ae8e3c7..0a95370 100644
--- a/include/asm-arm/arch-mxc/mx31.h
+++ b/include/asm-arm/arch-mxc/mx31.h
@@ -1,5 +1,5 @@
 /*
- * Copyright 2004-2007 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright 2004-2008 Freescale Semiconductor, Inc. All Rights Reserved.
  */
 
 /*
@@ -194,6 +194,7 @@
 #define MXC_DMA_CHANNEL_ATA_RX  MXC_DMA_DYNAMIC_CHANNEL
 #define MXC_DMA_CHANNEL_ATA_TX  MXC_DMA_DYNAMIC_CHANNEL
 #define MXC_DMA_CHANNEL_MEMORY  MXC_DMA_DYNAMIC_CHANNEL
+#define MXC_DMA_CHANNEL_FIFO_MEMORY  MXC_DMA_DYNAMIC_CHANNEL
 
 /*
  * AIPS 2
@@ -457,4 +458,4 @@
 #define NFMS_NF_DWIDTH 		31
 #define NFMS_NF_PG_SZ 		30
 
-#endif			/*  __ASM_ARCH_MXC_MX31_H__ */
+#endif				/*  __ASM_ARCH_MXC_MX31_H__ */
diff --git a/include/asm-arm/arch-mxc/sdma.h b/include/asm-arm/arch-mxc/sdma.h
index f259008..ae848f1 100644
--- a/include/asm-arm/arch-mxc/sdma.h
+++ b/include/asm-arm/arch-mxc/sdma.h
@@ -1,6 +1,6 @@
 
 /*
- * Copyright 2004-2007 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright 2004-2008 Freescale Semiconductor, Inc. All Rights Reserved.
  */
 
 /*
@@ -53,6 +53,8 @@
 #define MAX_DMA_CHANNELS 0
 #endif
 
+#define MXC_FIFO_MEM_DEST_FIXED   0x1
+#define MXC_FIFO_MEM_SRC_FIXED    0x2
 /*!
  * This enumerates  transfer types
  */
@@ -97,7 +99,8 @@ typedef enum {
 	EXT,			/*!< External peripheral */
 	MSHC,			/*!< Memory Stick Host Controller */
 	DSP,			/*!< DSP */
-	MEMORY			/*!< Memory */
+	MEMORY,			/*!< Memory */
+	FIFO_MEMORY		/*!< FIFO type Memory */
 } sdma_periphT;
 
 #ifndef TRANSFER_32BIT
@@ -190,6 +193,8 @@ typedef struct {
 	int mxc_sdma_ap_2_ap_addr;
 	/*! address of ap_2_bp script */
 	int mxc_sdma_ap_2_bp_addr;
+	/*! address of ap_2_ap_fixed script */
+	int mxc_sdma_ap_2_ap_fixed_addr;
 	/*! address of bp_2_ap script */
 	int mxc_sdma_bp_2_ap_addr;
 	/*! address of loopback_on_dsp_side script */
-- 
1.5.4.4

